### PR TITLE
Add clang-format and clang-tidy checks into Jenkins Pipeline

### PR DESCRIPTION
**Pull request by yiqian1**

### CODE REVIEW

File: mlir/utils/jenkins/Jenkinsfile.downstream [hunk 1]
```diff
@@ -43,6 +43,13 @@ pipeline {
 
                     # build LLVM / MLR and run tests
                     cmake --build . --target check-mlir
+
+	     	    # premerge checks
+                    cd ..
+                    if [ ! -f ./compile_commands.json ]; then
+                       ln -s build/compile_commands.json compile_commands.json
+                    fi
+		    python3.7 ./mlir/utils/jenkins/premerge-checks.py
```
Comment (jerryyin) [inline mlir/utils/jenkins/Jenkinsfile.downstream]: This needs to be an independent stage. It needs to run before the actual check-mlir. Also this should be an independent github check item as in acceptance criteria of the user story:
> The pre-merge task, should list as a separate item of check in GitHub and block further checks, if not passed.

File: mlir/utils/jenkins/clang-format.ignore [hunk 2]
```diff
@@ -0,0 +1 @@
+# Patterns for clang-format to ignore.
```
Comment (jerryyin) [inline mlir/utils/jenkins/clang-format.ignore]: Add a comment to point to https://github.com/google/llvm-premerge-checks/blob/master/scripts/clang-format.ignore

Also, you probably need to include the `**/test` pattern so that it does not target test files.

File: mlir/utils/jenkins/clang-tidy.ignore [hunk 3]
```diff
@@ -0,0 +1,5 @@
+# Files to be ignored by clang-tidy. Will not appear in short report and inline comments.
```
Comment (jerryyin) [inline mlir/utils/jenkins/clang-tidy.ignore]: Add a comment to point to https://github.com/google/llvm-premerge-checks/blob/master/scripts/clang-tidy.ignore

File: mlir/utils/jenkins/premerge-checks.py [hunk 4]
```diff
@@ -0,0 +1,104 @@
+#!/usr/bin/env python3
```
Comment (jerryyin) [inline mlir/utils/jenkins/premerge-checks.py]: Desirably, add a comment before each function referring to where to find it in the `premerge-checks` repo. Point them to `premerge_checks.py` / `clang_format_report.py` / `clang-tidy_report.py`.

Also, add a llvm license header section with a brief description of this file, and document the bash command to run for a individual file.

File: mlir/utils/jenkins/premerge-checks.py [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+import argparse
+import logging
+import os
+import re
+import sys
+import subprocess
+from typing import Tuple, Optional
+import pathspec
+import unidiff
```
Comment (jerryyin) [inline mlir/utils/jenkins/premerge-checks.py]: You may need to `pip3 install` so that those packages are available. If it is already there, feel free to ignore this.

File: mlir/utils/jenkins/clang-format.ignore (hunk 2 above)
Comment (jerryyin) [inline mlir/utils/jenkins/clang-format.ignore]: Now on a second look, it probably makes more sense to move scripts and ignore files under a dedicated folder, for example, `mlir/utils/jenkins/static-checks/`

File: mlir/utils/jenkins/Jenkinsfile.downstream [hunk 6]
```diff
@@ -43,6 +43,13 @@ pipeline {
 
                     # build LLVM / MLR and run tests
                     cmake --build . --target check-mlir
+
+	     	    # premerge checks
+                    cd ..
+                    if [ ! -f ./compile_commands.json ]; then
+                       ln -s build/compile_commands.json compile_commands.json
+                    fi
```
Comment (jerryyin) [inline mlir/utils/jenkins/Jenkinsfile.downstream:58]: Why do you need this symlink?

File: mlir/utils/jenkins/Jenkinsfile.downstream [hunk 7]
```diff
@@ -18,7 +18,7 @@ pipeline {
                 sh '''
```
Comment (jerryyin) [inline mlir/utils/jenkins/Jenkinsfile.downstream]: A high level question, do you need any of those [lines](https://github.com/google/llvm-premerge-checks/blob/master/containers/base-debian/Dockerfile#L41-L48)? I feel like you'd need to install following [this line](https://github.com/google/llvm-premerge-checks/blob/master/containers/base-debian/Dockerfile#L20) as well.

**Review by jerryyin — CHANGES_REQUESTED** [review]

File: mlir/utils/jenkins/Jenkinsfile.downstream (hunk 7 above)
Comment (yiqian1) [reply mlir/utils/jenkins/Jenkinsfile.downstream]: We don't need clang-10. Our default clang is clang-11.

File: mlir/utils/jenkins/Jenkinsfile.downstream (hunk 6 above)
Comment (yiqian1) [reply mlir/utils/jenkins/Jenkinsfile.downstream:58]: Clang-tidy needs this file and it has to be at the llvm-project-mlir directory.

Comment (jerryyin) [reply mlir/utils/jenkins/Jenkinsfile.downstream:58]: Is this because of `clang-tidy` being not invoked from the `build` directory? Can it be invoked in `build` directory?

Comment (yiqian1) [reply mlir/utils/jenkins/Jenkinsfile.downstream:58]: No, it gave file-not-found kind of errors. And Gibrater does the symlink as well.

File: mlir/utils/jenkins/Jenkinsfile.downstream (hunk 1 above)
Comment (yiqian1) [reply mlir/utils/jenkins/Jenkinsfile.downstream]: As it needs compile_commands.json, it should after cmake config and before cmake --build. I mean comile_commands.json is generated by cmake config, right?

Comment (jerryyin) [reply mlir/utils/jenkins/Jenkinsfile.downstream]: Sure, configuration goes before the actual `premerge-checks.py`, but `check-mlir` should be invoked last. In my opinion, `check-mlir` should be blocked until static checks all pass.

File: mlir/utils/jenkins/Jenkinsfile.downstream (hunk 6 above)
Comment (jerryyin) [reply mlir/utils/jenkins/Jenkinsfile.downstream:58]: Sounds good.

File: mlir/utils/jenkins/Jenkinsfile.downstream [hunk 8]
```diff
@@ -18,11 +18,16 @@ pipeline {
...
                     # install pkg dependencies
                     sudo apt update
-                    sudo apt install -y ninja-build wget git python cmake
+                    sudo apt install -y ninja-build wget git python3.7 cmake python3-pip
+		    if [ -f /usr/bin/python ]; then
+                      	/bin/rm /usr/bin/python
+                    fi 
+		    ln -s /usr/bin/python3.7 /usr/bin/python
```
Comment (jerryyin) [inline mlir/utils/jenkins/Jenkinsfile.downstream]: This is wrong. Directly removing the binary is very hacky and can confuse the package manager.

Choose one of the following ways:
 - `alias python=python3`
 - Remove the python package before installation of python3.7

File: mlir/utils/jenkins/Jenkinsfile.downstream [hunk 9]
```diff
@@ -41,8 +46,27 @@ pipeline {
...
                         -DMLIR_MIOPEN_DRIVER_ENABLED=1
 
-                    # build LLVM / MLR and run tests
-                    cmake --build . --target check-mlir
+	     	    # premerge checks
+                    cd ..
+                    if [ ! -f ./compile_commands.json ]; then
+                       ln -s build/compile_commands.json compile_commands.json
+                    fi
+		    python3.7 ./mlir/utils/jenkins/static-checks/premerge-checks.py
```
Comment (jerryyin) [inline mlir/utils/jenkins/Jenkinsfile.downstream]: Why does the CI show the following:
```
+ python3.7 ./mlir/utils/jenkins/static-checks/premerge-checks.py
no modified files to format
```
Is it because you are changing the jenkins folder, which clang static checks doesn't care about? If that's the case, we might need a case where manually introduce a change that can make clang static checks to break. Just to make sure the static check can actually work.

File: mlir/utils/jenkins/Jenkinsfile.downstream [hunk 10]
```diff
@@ -18,11 +18,16 @@ pipeline {
...
+		    if [ -f /usr/bin/python ]; then
+                      	/bin/rm /usr/bin/python
+                    fi 
+		    ln -s /usr/bin/python3.7 /usr/bin/python
+	 	    python3.7 -m pip install pathspec unidiff
                 '''
             }
         }
-        stage('Test') {
+        stage('Static Test') {
```
Comment (jerryyin) [inline mlir/utils/jenkins/Jenkinsfile.downstream:33]: Thanks for making static test as an independent stage. Though github still shows everything as a single item. Maybe a good time to ask the MIOpen channel about how they did it?

File: mlir/utils/jenkins/Jenkinsfile.downstream (hunk 1 above)
Comment (yiqian1) [reply mlir/utils/jenkins/Jenkinsfile.downstream]: It seems that showing build status for each stage is done via a Jenkins plugin github-autostatus. There is no need to change the jenkinsfile. However I couldn't find this plugin in our Jenkins system.

File: mlir/utils/jenkins/static-checks/premerge-checks.py [hunk 11]
```diff
@@ -0,0 +1,130 @@
+#!/usr/bin/env python3
```
Comment (jerryyin) [inline mlir/utils/jenkins/static-checks/premerge-checks.py:1]: The formatting of this file is wired. You have the tabspace to be 2/3/4. The python files under llvm folder are all 2 space by default. I suggest you to run static checks on this file yourself to fix the formatting issues.

File: mlir/utils/jenkins/Jenkinsfile.downstream (hunk 8 above)
Comment (yiqian1) [reply mlir/utils/jenkins/Jenkinsfile.downstream]: I removed a symlink

Comment (jerryyin) [reply mlir/utils/jenkins/Jenkinsfile.downstream]: Same concept applies for symlink. The link must be created by **some package**, likely python. Now you are removing a simlink of the package without actually removing the entire package. This creates an edge case for package manager to handle.

File: mlir/utils/jenkins/Jenkinsfile.downstream (hunk 10 above)
Comment (yiqian1) [reply mlir/utils/jenkins/Jenkinsfile.downstream:33]: It seems that showing build status for each stage is done via a Jenkins plugin github-autostatus. There is no need to change the jenkinsfile. However I couldn't find this plugin in our Jenkins system.

File: mlir/utils/jenkins/Jenkinsfile.downstream (hunk 1 above)
Comment (jerryyin) [reply mlir/utils/jenkins/Jenkinsfile.downstream]: That's right I can confirm the available plugin is [autostatus](https://plugins.jenkins.io/github-autostatus/). Looks like it requires a newer jenkins version. I think we should do the upgrade, and this should be orthogonal to this PR. 

Can you create a new issue in [`tensorflow-internal`](https://github.com/ROCmSoftwarePlatform/tensorflow-internal/issues). Describe about the background there. In the weekly meeting, ask the teams opinion about upgrade our Jenkins (this/next week, depends on your preference.)

File: mlir/utils/jenkins/Jenkinsfile.downstream (hunk 9 above)
Comment (jerryyin) [reply mlir/utils/jenkins/Jenkinsfile.downstream]: It would be better if we can introduce static checks to go through the files under jenkins folder.

File: mlir/utils/jenkins/Jenkinsfile.downstream (hunk 1 above)
Comment (yiqian1) [reply mlir/utils/jenkins/Jenkinsfile.downstream]: I've created one: https://github.com/ROCmSoftwarePlatform/tensorflow-internal/issues/83

Please let me know if I did it right or not.

Comment (jerryyin) [reply mlir/utils/jenkins/Jenkinsfile.downstream]: Yes, you created one correctly. Could you put more details in that? TF people won't be able to understand just with the description you had. For example, 
 - this is the ask from MLIR
 - Reference back to the discussion here
 - Give a link and name to the Jenkins plugin needed
 - Mention what the current Jenkins version is, and what is the minimum required version for plugin
 - Give a little more details on what the plugin dose and attach a image to show the end result

File: mlir/utils/jenkins/Jenkinsfile.downstream (hunk 9 above)
Comment (yiqian1) [reply mlir/utils/jenkins/Jenkinsfile.downstream]: By default, clang-format formats C/C++/Java/JavaScript/Objective-C/Protobuf/C# code and skips other types of files. I tried to ran clang-format on premerge-checks.py, it generates weird and incorrect results. For example, it added spaces between words in #!/usr/bin/env phthon3. Looks like it treated this line as a normal comment. And it didn't fix inconsistent indents. I think applying static checks to jenkins folder needs
 better understanding of how clang-format and clang-tidy work and should belong to another review.

Comment (yiqian1) [reply mlir/utils/jenkins/Jenkinsfile.downstream]: I have manually changed a mlir source file and verified that clang-format and clang-tidy would stop and issue an exit code of 1, as expected.

Comment (jerryyin) [reply mlir/utils/jenkins/Jenkinsfile.downstream]: Thanks for helping to figure out the file type part. That sounds reasonable to skip the python file check as upstream do.

Regarding on manually introducing the change, if you don't think that's necessary, then could you keep an closer eye on the CI state after your submission? Just to make sure everything works as expected.

File: mlir/utils/jenkins/static-checks/premerge-checks.py (hunk 11 above)
Comment (jerryyin) [reply mlir/utils/jenkins/static-checks/premerge-checks.py:1]: I'm referencing to the formatting of [this file](https://github.com/ROCmSoftwarePlatform/llvm-project-mlir/blob/miopen-dialect/mlir/utils/generate-test-checks.py). I believe `autopep8` can do it for you.

File: mlir/utils/jenkins/Jenkinsfile.downstream (hunk 9 above)
Comment (yiqian1) [reply mlir/utils/jenkins/Jenkinsfile.downstream]: Sure. I will.

File: mlir/utils/jenkins/Jenkinsfile.downstream (hunk 8 above)
Comment (yiqian1) [reply mlir/utils/jenkins/Jenkinsfile.downstream]: This should be Ok. I didn't install python. (Docker automatically installed python3.6. ) Using an alias or removing the python package wouldn't help because python is not installed.  Alternatively, I may be able to use the old version of subprocess() to avoid installation of python3.7, as the reason why I wanted to have python3.7 was to make two subprocess() calls happy.  I tried the old version of subprocess() and that worked in my local machine, but the build failed at Jenkins. I will take a look.

File: mlir/utils/jenkins/Jenkinsfile.downstream [hunk 12]
```diff
@@ -18,11 +18,13 @@ pipeline {
                 sh '''
                     # install pkg dependencies
                     sudo apt update
-                    sudo apt install -y ninja-build wget git python cmake
+                    sudo apt install -y ninja-build wget git python cmake python3-pip
+                    pip3 install pathspec unidiff
                 '''
+                sh "PATH=$PATH:/opt/rocm/llvm/bin/"
```
Comment (jerryyin) [inline mlir/utils/jenkins/Jenkinsfile.downstream]: Why is this necessary?

File: mlir/utils/jenkins/Jenkinsfile.downstream [hunk 13]
```diff
@@ -41,8 +43,27 @@ pipeline {
                         -DMLIR_ROCM_RUNNER_ENABLED=1 \
                         -DMLIR_MIOPEN_DRIVER_ENABLED=1
 
-                    # build LLVM / MLR and run tests
-                    cmake --build . --target check-mlir
+	     	    # premerge checks
+                    cd ..
+                    if [ ! -f ./compile_commands.json ]; then
+                       ln -s build/compile_commands.json compile_commands.json
+                    fi
```
Comment (jerryyin) [inline mlir/utils/jenkins/Jenkinsfile.downstream:58]: Can we get rid of this `cd..`? The `ln -s` can handle a different path.

File: mlir/utils/jenkins/Jenkinsfile.downstream [hunk 14]
```diff
@@ -41,8 +43,27 @@ pipeline {
...
+		    python3 ./mlir/utils/jenkins/static-checks/premerge-checks.py
+
+		    if [ $? != 0 ]; then
+			exit 1
+		    fi
```
Comment (jerryyin) [inline mlir/utils/jenkins/Jenkinsfile.downstream]: I think Jenkins does this by default, does it?

Comment (yiqian1) [reply mlir/utils/jenkins/Jenkinsfile.downstream]: I think by default Jenkins will exit the current stage and continue to the next stage.

File: mlir/utils/jenkins/Jenkinsfile.downstream (hunk 13 above)
Comment (yiqian1) [reply mlir/utils/jenkins/Jenkinsfile.downstream:58]: No. I cannot invoke the static checks in the build directory. That would cause errors. I'm not exactly sure why.

File: mlir/utils/jenkins/Jenkinsfile.downstream (hunk 14 above)
Comment (jerryyin) [reply mlir/utils/jenkins/Jenkinsfile.downstream]: That's probably not true. From my experience with MIOpen CI, they always stop when one stage has a failure, yet their [Jenkinsfile](https://github.com/ROCmSoftwarePlatform/MIOpen/blob/develop/Jenkinsfile) does not come with any `exit 1`.

File: mlir/utils/jenkins/Jenkinsfile.downstream (hunk 12 above)
Comment (yiqian1) [reply mlir/utils/jenkins/Jenkinsfile.downstream]: git-clang-format needs this path to find clang-format.

File: mlir/utils/jenkins/Jenkinsfile.downstream (hunk 14 above)
Comment (yiqian1) [reply mlir/utils/jenkins/Jenkinsfile.downstream]: Their code catches exceptions. Maybe that is how it exits.

Comment (yiqian1) [reply mlir/utils/jenkins/Jenkinsfile.downstream]: Do you know whom in MIOpen group I could contact with to ask this question?

Comment (jerryyin) [reply mlir/utils/jenkins/Jenkinsfile.downstream]: @daniellowell is the expert in this. Is this explicit `exit 1` necessary in the pipeline for jenkins to abort execution?

File: mlir/utils/jenkins/Jenkinsfile.downstream (hunk 13 above)
Comment (jerryyin) [reply mlir/utils/jenkins/Jenkinsfile.downstream:58]: Could you add a comment saying that the static checks must be carried out in the root directory?

File: mlir/utils/jenkins/Jenkinsfile.downstream (hunk 12 above)
Comment (jerryyin) [reply mlir/utils/jenkins/Jenkinsfile.downstream]: Could you add a comment and be explicit about this?

File: mlir/utils/jenkins/Jenkinsfile.downstream (hunk 14 above)
Comment (jerryyin) [reply mlir/utils/jenkins/Jenkinsfile.downstream]: @yiqian1  Now instead of waiting for the reply, why not push for an experimental commit and see if Jenkins behaves as expected? You can easily `reset HEAD` to original once you figured it out.

**Review by jerryyin — APPROVED** [review]
Thanks, LGTM.
